File: CSharp/Supermarket.Specs/Tools/ReceiptPrinter.cs
using System;
using System.Globalization;
using System.Text;
using Supermarket.Model.Checkout;

namespace Supermarket.Specs.Tools
{
    public class ReceiptPrinter
    {
        public string Print(Receipt receipt)
        {
            if (receipt == null)
                throw new ArgumentNullException(nameof(receipt));

            var sb = new StringBuilder();

            AppendBookingLines(receipt, sb);
            AppendTotal(receipt, sb);
            AppendTaxLines(receipt, sb);

            return sb.ToString();
        }

        private static void AppendTaxLines(Receipt receipt, StringBuilder sb)
        {
            foreach (var line in receipt.TaxLines)
            {
                sb.AppendLine($"Incl. {line.Rate / 100:P} VAT {DecimalToString(line.Amount)}");
            }
        }

        private static void AppendTotal(Receipt receipt, StringBuilder sb)
        {
            sb.AppendLine("-----------------------");
            sb.AppendLine($"Total {DecimalToString(receipt.Total)}");
        }

        private static void AppendBookingLines(Receipt receipt, StringBuilder sb)
        {
            foreach (var line in receipt.Bookings)
            {
                sb.AppendLine($"{line.Text} {DecimalToString(line.Price)}");
            }
        }

        private static string DecimalToString(decimal dec)
        {
            return dec.ToString("C", new CultureInfo("en-US"));
        }
    }
}

File: java/src/test/java/supermarket/ReceiptPrinter.java
package supermarket;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.math.BigDecimal;
import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.util.Locale;

import supermarket.model.checkout.BookingLine;
import supermarket.model.checkout.Receipt;
import supermarket.model.checkout.TaxLine;

public class ReceiptPrinter {

	public String print(Receipt receipt) {
		if (receipt == null) {
			throw new IllegalArgumentException();
		}

		StringWriter stringWriter = new StringWriter();
		PrintWriter printWriter = new PrintWriter(stringWriter);

		appendBookingLines(receipt, printWriter);
		appendTotal(receipt, printWriter);
		appendTaxLines(receipt, printWriter);

		return printWriter.toString();
	}

	private void appendTaxLines(Receipt receipt, PrintWriter sb) {
		for (TaxLine taxline : receipt.getTaxLines()) {
			sb.println(format(taxline));
		}
	}

	private void appendTotal(Receipt receipt, PrintWriter sb) {
		sb.println(String.format("%-33s", " ").replace(' ', '-'));
		sb.append(String.format("        %-25s %s", "Total", formatCurrency(receipt.getTotal())));
	}

	private void appendBookingLines(Receipt receipt, PrintWriter sb) {
		for (BookingLine booking : receipt.getBookings()) {
			sb.println(format(booking));
		}
	}

	public String formatCurrency(BigDecimal amount) {
		NumberFormat numberFormat = DecimalFormat.getInstance(Locale.ENGLISH);
		numberFormat.setMinimumFractionDigits(2);
		numberFormat.setMaximumFractionDigits(2);
		return String.format("%s€", numberFormat.format(amount));
	}

	private String formatPercentage(int rate) {
		NumberFormat numberFormat = NumberFormat.getPercentInstance();
		return numberFormat.format((double) rate / 100);
	}

	public String format(TaxLine taxLine) {
		return String.format("        Incl. %s VAT            %s", formatPercentage(taxLine.getRate()),
				formatCurrency(taxLine.getAmount()));
	}

	public String format(BookingLine booking) {
		return String.format("        %-25s %s", booking.getText(), formatCurrency(booking.getPrice()));
	}

}
